### PR TITLE
[Merged by Bors] - Reduce the use of atomics in the render phase

### PR DESCRIPTION
**Pull request by james7132**
# Objective
Speed up the render phase of rendering. An extension of #6885.

`SystemState::get` increments the `World`'s change tick atomically every time it's called. This is notably more expensive than a unsynchronized increment, even without contention. It also updates the archetypes, even when there has been nothing to update when it's called repeatedly.

## Solution
Piggyback off of #6885. Split `SystemState::validate_world_and_update_archetypes` into `SystemState::validate_world` and `SystemState::update_archetypes`, and make the later `pub`. Then create safe variants of `SystemState::get_unchecked_manual` that still validate the `World` but do not update archetypes and do not increment the change tick using `World::read_change_tick` and `World::change_tick`. Update `RenderCommandState` to call `SystemState::update_archetypes` in `Draw::prepare` and `SystemState::get_manual` in `Draw::draw`.

## Performance
There's a slight perf benefit (~2%) for `main_opaque_pass_3d` on `many_foxes` (340.39 us -> 333.32 us)

image

## Alternatives
We can change `SystemState::get` to not increment the `World`'s change tick. Though this would still put updating the archetypes and an atomic read on the hot-path.

---

## Changelog
Added: `SystemState::get_manual`
Added: `SystemState::get_manual_mut`
Added: `SystemState::update_archetypes`

### CODE REVIEW

**Review by Guvante — APPROVED** [review]
I am uncertain about the get_manual nomenclature but it matches other call sites so makes sense.

**Review by kurtkuehnert — APPROVED** [review]
This change makes sense to me and is in line with the changes of #6885, but I am no ECS expert.

File: crates/bevy_ecs/src/system/function_system.rs [hunk 1]
```diff
@@ -196,8 +198,13 @@ impl<Param: SystemParam> SystemState<Param> {
...
     }
 
-    fn validate_world_and_update_archetypes(&mut self, world: &World) {
+    #[inline]
+    fn validate_world(&mut self, world: &World) {
         assert!(self.matches_world(world), "Encountered a mismatched World. A SystemState cannot be used with Worlds other than the one it was created with.");
+    }
+
+    #[inline]
+    pub fn update_archetypes(&mut self, world: &World) {
```
Comment (JoJoJet) [inline crates/bevy_ecs/src/system/function_system.rs:207]: This should have docs explaining when this needs to be called manually, and when the user doesn't need to worry about it.

**Review by JoJoJet — CHANGES_REQUESTED** [review]
LGTM overall, there's just one nit.

File: crates/bevy_ecs/src/system/function_system.rs [hunk 2]
```diff
@@ -211,6 +218,42 @@ impl<Param: SystemParam> SystemState<Param> {
         }
     }
 
+    /// Retrieve the [`SystemParam`] values. This can only be called when all parameters are read-only.
+    /// This will not update archetypes automatically nor increment the world's change tick.
```
Comment (mockersf) [inline crates/bevy_ecs/src/system/function_system.rs:222]: this line doesn't feel accurate: it's not the archetypes that need to be updated, it's the state view of the world archetypes

File: crates/bevy_ecs/src/system/function_system.rs [hunk 3]
```diff
@@ -196,8 +198,13 @@ impl<Param: SystemParam> SystemState<Param> {
         self.world_id == world.id()
     }
 
-    fn validate_world_and_update_archetypes(&mut self, world: &World) {
+    #[inline]
+    fn validate_world(&mut self, world: &World) {
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/system/function_system.rs:202]:
```suggestion
    /// Asserts that the [`SystemState`] matches the provided [`World`].
    fn validate_world(&mut self, world: &World) {
```

File: crates/bevy_ecs/src/system/function_system.rs (hunk 1 above)
Comment (alice-i-cecile) [reply crates/bevy_ecs/src/system/function_system.rs:207]: And what this method generally *does*.

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]
Agreed with the review comments: the docs need to be improved before I'd be comfortable with merging it.

**Comment by kurtkuehnert** [comment]
Is there a way we can make `SystemStare::get_manual` take an immutable reference to self? Similar to `QueryState::get_manual`.
This would make 'RenderCommandState::render` immutable as well.

**Comment by JoJoJet** [comment]
> Is there a way we can make `SystemStare::get_manual` take an immutable reference to self? Similar to `QueryState::get_manual`.

At first glance, this seems very difficult to generalize. We'd have to change `SystemParam::get_param` to take an immutable refence to the state, which would break params that need to modify their state such as `Commands` or `Local` or `EventReader`.

One way of doing this would be to add a new trait for `SystemParam`s that do not mutably access their state in `get_param`, and then add `SystemState::get_readonly` which is constrained to using this trait. This would allow *most* `SystemParam`s to be used with a read-only `SystemState`, while excluding the param types that don't make sense. I am not exactly sure what this would look like, though.

**Review by JoJoJet — APPROVED** [review]

**Comment by kurtkuehnert** [comment]
I think making `SystemState::get_manual` immutable would allow us to get rid of the `RwLock` in the `DrawFunctions` struct. Although I do not believe that this will result in any significant performance improvements.  
Thus, I do not think that is worth the effort.
Anyways, thanks for the reply. 😄

File: crates/bevy_ecs/src/system/function_system.rs [hunk 4]
```diff
@@ -212,6 +226,43 @@ impl<Param: SystemParam> SystemState<Param> {
         }
     }
 
+    /// Retrieve the [`SystemParam`] values. This can only be called when all parameters are read-only.
+    /// This will not update the state's view of the world's archetypes automatically nor increment the
+    /// world's change tick.
+    ///
+    /// For this to return accurate results, ensure [`SystemState::update_archetypes`] is called before this
+    /// function.
```
[the remark this answers is not in the document]
Comment (james7132) [reply crates/bevy_ecs/src/system/function_system.rs:234]: New archetypes are ignored so it effectively ignores them. Queries, for example, will just not return the target entity in those new archetypes. The prior archetypes are still accessed correctly, so this shouldn't cause any UB if not updated.

**Comment by james7132** [comment]
> I think making `SystemState::get_manual` immutable would allow us to get rid of the `RwLock` in the `DrawFunctions` struct. Although I do not believe that this will result in any significant performance improvements.  
> Thus, I do not think that is worth the effort.

It might be. If we ever extract the render phases into their own systems, the write lock on DrawFunctions means we can't parallelize properly on a per-view basis (via Query::par_for_each).

**Review by alice-i-cecile — APPROVED** [review]
bors r+

**Pull request closed without merging**